### DISCUSSION TITLE
Execute an action onDisappear

### DISCUSSION THREAD

**Discussion by Alexminator99** [pointfreeco/swift-composable-architecture#2891] [Q&A] [answered]
Hi there community. I've been reading a lot about cancellation and onDisappear action execution with TCA but I haven't  found yet one that explains what to do if I need to close some resource.

Lets take the following example from StarMicronicsSDK:
- They have printers that can be opened and closed.
- Both operations are async so is easy to implement a dependency that either starts monitoring and returns some state or stop monitoring a device
- Right now I'm running starMonitoring action using .task modifier and a cancellable effect.

So, whenever the view disappears I need to stop monitoring the device and sending an action when event .onDisappear triggers may lead to a warning/error from TCA that a state was executed when child was nil. 

The app right now displays a list of printers you can pick from and navigates to a detail view where it starts monitoring as I mentioned earlier. 

Here comes my question, how can I stop monitoring when user goes back (they pop from the stack) or when the view disappears?

Here is some code I can share to give you guys a hint:

## Reducer
```swift
case .startMonitoring:
 return .concatenate(
       .send(.detectPaperWidth),
        monitor.startMonitoring(state.printer.id)
   )
   .cancellable(id: CancelID.cancelScanning, cancelInFlight: true)
```

## View
```swift
.task {
   await self.store.send(.startMonitoring).finish()
}
.onDisappear {
  self.store.send(.stopMonitoring) // This is NOT the way
}
```

## Parent Reducer logic:
```swift
...
var path = StackState<HardwareDetailFeature.State>()
...
case path(StackAction<HardwareDetailFeature.State, HardwareDetailFeature.Action>)
...
.forEach(\.path, action: \.path) {
     HardwareDetailFeature()
}
```

## Parent View
```swift
NavigationStack(path: $store.scope(state: \.path, action: \.path)) {
....
NavigationLink(state: HardwareDetailFeature.State(
                                    printer: printer,
                                    printerIsConnected: true
)) {
    EmptyView() // Just a hack to avoid the swiftui chevron and use my own
 }
 ```
 
 ## System:
 - Xcode 15.2
 - Swift 5.9
 - TCA 1.9.0

**Comment by mbrandonw** [comment]
> Here comes my question, how can I stop monitoring when user goes back (they pop from the stack) or when the view disappears?

This happens automatically when using `forEach`. All effects are cancelled for the feature that is popped off the stack. There is no need to send an `onDisappear` action and cancel effects manually.

**Reply by Alexminator99** [reply]
> > Here comes my question, how can I stop monitoring when user goes back (they pop from the stack) or when the view disappears?
> 
> 
> 
> This happens automatically when using `forEach`. All effects are cancelled for the feature that is popped off the stack. There is no need to send an `onDisappear` action and cancel effects manually.

Mmm the effect may be cancellable but I need to specifically call stop monitoring in order to properly close connection to the printer, otherwise printer will remain open and further operations may throw device already opened due to the fact that despite for each cancelled the effect the printer is still open. Am I missing something here 😅?

**Reply by mbrandonw** [reply, marked as answer]
Yes, the effect being cancelled is all you need. If your effect is constructed using a Combine publisher then you can use something like `.handleEvents` or `AnyCancellable` to detect cancellation and clean up resources. Or if you are use Swift's async tools then you can use something like `onTermination` from `AsyncStream.Continuation` or `withCancellationHandler`. 

Essentially it's your job to tap into cancellation in the effect and do the clean up. All the library needs to do is cancel the effect.

**Reply by Alexminator99** [reply]
> Yes, the effect being cancelled is all you need. If your effect is constructed using a Combine publisher then you can use something like `.handleEvents` or `AnyCancellable` to detect cancellation and clean up resources. Or if you are use Swift's async tools then you can use something like `onTermination` from `AsyncStream.Continuation` or `withCancellationHandler`. 
> 
> 
> 
> Essentially it's your job to tap into cancellation in the effect and do the clean up. All the library needs to do is cancel the effect.

Thanks I will try this but I get the point you were trying to make. Thanks 🙏

**Reply by Alexminator99** [reply]
Your solution worked like a charm @mbrandonw thks!!!

```swift
func startMonitoring(identifier: String) async -> AsyncStream<MonitorAction> {
        AsyncStream { continuation in
            self.monitorContinuation = continuation
            
            Task {
                await self.monitor(for: identifier)
            }
            
            self.monitorContinuation?.onTermination =  { @Sendable _ in
                Task {
                    await self.stopMonitoring()
                }
            }
        }
    }
```
